### ISSUE TITLE
Support for mutually conflicting options

### ISSUE THREAD

**Issue by kravemir** [remkop/picocli#414]
I want to offer multiple ways to define parameters for operation. For example parameters are following: paper size, paper margins, and various other parameters. And, there should be at least two ways to define them:

* `--paper-definition-json FILE`, or
* `--paper-definition-xml FILE`, or
* `--paper-size VALUE...`, `--paper-margins VALUE...`, `--paper-param-a ...`, `--paper-param-b`, and so on...

These options are reused in same mix-in in various multiple sub-commands. Is there any (considered) support for such cases? Or, should I create own mixin content validation mechanism, and call it inside each sub-command using the mixin?

While I was writing this, I thought, that mixin `PaperMixin` has got `Paper getPaperDefinition()` method,.. validation can happen there, no addition code needed inside each sub-command.

But, as I have already wrote this, .. it might be good idea to implement automatic validation of conflicting options, as it could be reusable in other applications.

**Comment by remkop** [comment]
Yes, picocli currently does not do validation (other then when converting types).
When I need validation I usually create a separate class for this so I can reuse it from multiple commands.
Does that answer your question?

**Comment by kravemir** [comment]
And, is there any standard (reusable) way to output errors of input arguments from custom validation logic? Also, what approach would you suggest for handling this, and indication about this in `--help`?

**Comment by remkop** [comment]
Hmm, the official standard for validation is [jsr 303](https://beanvalidation.org), but that may be a bit heavy weight for a stand-alone application. 

There’s several ways to do this. You can throw an exception or you can have a `validate` method (it’s a good idea to allow returning multiple error messages from this method). Or you can invent your own annotations and inspect those at runtime. 

Generally it’s a good idea to keep it simple though.

**Comment by kravemir** [comment]
So, I guess, there's not gonna be any support for `InvalidCommandlineArgumentException` with properties `argument` and `reason`, in `picocli`? Which would have been caught and outputted pretty output with `exit(1)`?

**Comment by remkop** [comment]
Yes, at the moment I'm not planning to add support for validation to picocli.

**Comment by remkop** [comment]
What I’ll probably do is write an article using picocli with JSR-380 Beans Validation.

**Comment by kravemir** [comment]
@remkop, looking forward to see it :-)
